File: 41_rotting_oranges.js
/*
You are given an m x n grid where each cell can have one of three values:

0 representing an empty cell,
1 representing a fresh orange, or
2 representing a rotten orange.

Every minute, any fresh orange that is 4-directionally adjacent to a rotten 
orange becomes rotten.

Return the minimum number of minutes that must elapse until no cell has a 
fresh orange. If this is impossible, return -1.

1 <= m, n <= 10
grid[i][j] is 0, 1, or 2.
*/

// https://leetcode.com/problems/rotting-oranges/description/

var orangesRotting = function (grid) {
  let m = grid.length,
    n = grid[0].length,
    rotten = [],
    fresh = 0;
  // get all the rotten oranges and the number of fresh oranges
  for (let i = 0; i < m; i++) {
    for (let j = 0; j < n; j++) {
      if (grid[i][j] === 2) rotten.push([i, j, 0]);
      else if (grid[i][j] === 1) fresh += 1;
    }
  }
  // if there are no fresh oranges, return 0
  if (fresh === 0) return 0;
  // if there are no rotten oranges, return -1
  if (rotten.length === 0) return -1;
  // bfs
  let queue = [];
  for (let [i, j, time] of rotten) {
    queue.push([i, j, time]);
  }
  // bfs function: 1) update grid, 2) update fresh, 3) add to queue
  while (queue.length > 0) {
    let [i, j, time] = queue.shift();
    if (i < 0 || i >= m || j < 0 || j >= n || grid[i][j] < 2) continue;
    else {
      if (i + 1 < m && grid[i + 1][j] === 1) {
        grid[i + 1][j] = 2;
        fresh -= 1;
        if (fresh === 0) return time + 1;
        queue.push([i + 1, j, time + 1]);
      }
      if (i - 1 >= 0 && grid[i - 1][j] === 1) {
        grid[i - 1][j] = 2;
        fresh -= 1;
        if (fresh === 0) return time + 1;
        queue.push([i - 1, j, time + 1]);
      }
      if (j + 1 < n && grid[i][j + 1] === 1) {
        grid[i][j + 1] = 2;
        fresh -= 1;
        if (fresh === 0) return time + 1;
        queue.push([i, j + 1, time + 1]);
      }
      if (j - 1 >= 0 && grid[i][j - 1] === 1) {
        grid[i][j - 1] = 2;
        fresh -= 1;
        if (fresh === 0) return time + 1;
        queue.push([i, j - 1, time + 1]);
      }
    }
  }
  return -1;
};

File: 41_rotting_oranges.py
# /You are given an m x n grid where each cell can have one of three values:

# 0 representing an empty cell,
# 1 representing a fresh orange, or
# 2 representing a rotten orange.

# Every minute, any fresh orange that is 4-directionally adjacent to a rotten 
# orange becomes rotten.

# Return the minimum number of minutes that must elapse until no cell has a 
# fresh orange. If this is impossible, return -1.

# 1 <= m, n <= 10
# grid[i][j] is 0, 1, or 2.
# */

# https://leetcode.com/problems/rotting-oranges/

# BFS
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        import collections
        m, n = len(grid), len(grid[0])
        queue = collections.deque()
        fresh_cnt, minute = 0, 0
        
        # Step 1: Initialize the queue with all rotten oranges and count the fresh ones
        for i in range(m):
            for j in range(n):
                if grid[i][j] == 1:
                    fresh_cnt += 1
                # Add rotten oranges to the queue with time 0
                elif grid[i][j] == 2:
                    queue.append((i, j, 0)) 
        
        # Step 2: BFS to spread rot to neighboring fresh oranges
        while queue:
            i, j, minute = queue.popleft()
            
            # Visit the four neighboring cells
            for ni, nj in [(i + 1, j), (i - 1, j), (i, j + 1), (i, j - 1)]:
                if 0 <= ni < m and 0 <= nj < n and grid[ni][nj] == 1:
                    grid[ni][nj] = 2  # Mark fresh orange as rotten
                    fresh_cnt -= 1
                    queue.append((ni, nj, minute + 1))  # Spread the rot

        # Step 3: If there are no fresh oranges left, return the time; otherwise, return -1
        return minute if fresh_cnt == 0 else -1